### DISCUSSION TITLE
Replace nbdev?

### DISCUSSION THREAD

**Discussion by EtienneT** [marimo-team/marimo#2288] [Ideas]
Given that marimo is plain .py files, I was wondering if it could be used to replace something like [nbdev](https://nbdev.fast.ai/)?

It would be really nice to be able export specific functions defined in marimo cells.  Then you can define documentation, tests and example in other cells, testing and documenting the exported function.  This way you could import a marimo notebook in any other python file and use those python functions.

**Comment by akshayka** [comment]
> It would be really nice to be able export specific functions defined in marimo cells. 

This is on our roadmap, in particular we'll make it so that "pure" functions are serialized top-level into the notebook, and can be imported like any other function (`from notebook import my_function`) (cc @dmadisetti)

**Reply by EtienneT** [reply]
Awesome!  Is there a github issue we can track?

**Reply by akshayka** [reply]
Just made one: https://github.com/marimo-team/marimo/issues/2293

**Reply by dmadisetti** [reply]
Not familiar with nbdev, but exports are coming too!

**Reply by EtienneT** [reply]
> Not familiar with nbdev, but exports are coming too!

nbdev allows you to code your python library directly inside jupyter notebooks and then export only the functions your want to expose in your library.  You code the functions and then the documentation and tests inside the same notebook, but only export what you need for the library.  You can then generate your whole documentation directly from your notebooks.  It was made by Jeremy Howard who has one of the most popular machine learning course out there and he wrote the framework fast.ai with nbdev and also his last book (you can use nbdev to write books using notebooks).  The docs for fast.ai (and the python library itself) were generated using notebooks and nbdev: https://docs.fast.ai/.

Marimo being 100% python code for notebooks seems to be perfect to do something similar.

**Comment by leventov** [comment]
Something to consider:
 - Standard software "unit" tests,
 -  Data quality and statistical tests,
 - Various forms of "tests" appearing in ML and DS work (e.g., back-testing, cross-validation, sensitivity analysis, adversarial robustness, and many, many more kinds of tests), and
 - "Evals" for AI apps (or parts of them--LLM workflows and sub-routines)

I think all sufficiently different kinds of things to warrant different treatment. All these have specialised Python frameworks.

In nbdev itself, there is the concept of "test flags", and extra stuff people add to reduce the clutteredness of their code when these flags appear a lot: https://github.com/feynlee/nbdev-ignore.

One approach may be to automatically annotate various kinds of test cells by detecting the calls to specific functions from popular frameworks. Such as, if the cell contains calls to pytest, the cell is annotated as "test", no explicit tag/flag needed. Similar for code that calls to `openai/evals` by default annotated as "evals".
